### PR TITLE
doc: Gatekeeper policies as OCI image blog

### PR DESCRIPTION
**Pull request by mathieu-benoit**
Like discussed with @FeynmanZhou, proposing this blog going through an end-to-end scenario with bundling, testing and deploying OPA Gatekeeper policies as OCI image with ORAS, `gator` and GitOps.

Title: `Bundle, test and deploy Gatekeeper policies as OCI image`.

Thanks for your review, let me know if you have any feedback!

### CODE REVIEW

**Comment by FeynmanZhou** [comment]
Hi @mathieu-benoit, sorry for the late response. It's awesome to see your contributions to the ORAS blog. I just get back from the Chinese New Year holiday and continue to work on the ORAS project today. I will review it today. Thanks!

cc @oras-project/oras-client-release-managers for more attentions.

**Review by FeynmanZhou — APPROVED** [review]
LGTM.

cc @oras-project/oras-www-maintainers

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:1]: This title is a bit of a bag of words - maybe it's just me, but perhaps something shorter and more succinct?

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:5]: It would be appropriate to tell the reader what ORAS is - I can almost guarantee someone will come to the page and now know. E.g,. "ORAS "OCI Registry as Storage"

Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:5]:
```suggestion
[ORAS](https://oras.land/) is an important tool out there for working with OCI artifacts and OCI registries. As one of the users and advocates of ORAS, I witnessed the [growing trend in both user adoption and contributions in 2022](https://oras.land/blog/oras-looking-back-at-2022-and-forward-to-2023/). In this blog post I will share an end-to-end scenario with OPA Gatekeeper policies and ORAS, from including the steps from bundling to deployment.
```

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:7]:
```suggestion
Policies are rules expressed in YAML that not only afford meeting governance requirements, but also improve the security of Kubernetes workloads and clusters. Policy engines like [OPA Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/), [Kyverno](https://kyverno.io/) or even the new Kubernetes's [Validating Admission Policies](https://kubernetes.io/blog/2022/12/20/validating-admission-policies-alpha/) feature help write and enforce such policies. Once the policies are written, however, how do we easily and securely share them with different projects and teams? How do we deploy them across the fleet of clusters? How do we evaluate them as early as possible in CI/CD pipelines?
```

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:9]:
```suggestion
In this blog post we will demonstrate how to bundle and share Gatekeeper policies as an OCI image using the [ORAS](https://oras.land/cli/) command line client, how to evaluate Kubernetes manifests against this OCI image with the [`gator`](https://open-policy-agent.github.io/gatekeeper/website/docs/gator/) command line client, and how to deploy this OCI image in Kubernetes clusters in a GitOps way.
```

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:13]:
```suggestion
_While we use the [Google Artifact Registry](https://cloud.google.com/artifact-registry) as the OCI registry for this example, you can use [any registry supporting OCI artifacts](https://oras.land/implementors/#registries-supporting-oci-artifacts). For the GitOps tool, we are using the OSS project: [Config Sync](https://github.com/GoogleContainerTools/kpt-config-sync), and you can also use it as part of the [Anthos Config Management](https://cloud.google.com/anthos/config-management) service or even use other GitOps tools supporting OCI images like [FluxCD](https://fluxcd.io/flux/cheatsheets/oci-artifacts/)._
```

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:17]:
```suggestion
Let's create a [Gatekeeper policy](https://open-policy-agent.github.io/gatekeeper/website/docs/howto) composed by one `Constraint` and one `ConstraintTemplate` that will be leveraged throughout this blog. In this example, we are making sure that any non-system namespaces is leveraging the [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/) feature by having the appropriate label. 
```

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:21]:
```suggestion
mkdir -p policies
```

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:118]:
```suggestion
Let’s now locally test this `Namespace` against this policy with the `gator` CLI. This client makes it very convenient to test policies without any Kubernetes cluster!
```

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (vsoch) [inline docs/blog/gatekeeper-policies-as-oci-image.md:154]:
```suggestion
Let’s now locally test the `Namespace` created earlier against this policy as remote OCI image with the gator CLI. It is very convenient to share and evaluate your policies in different places (i.e. locally, during Continuous Integration pipelines, etc.)!
```

**Review by vsoch — COMMENTED** [review]
Thank you for writing this @mathieu-benoit ! I know Google has documentation standards and review - perhaps it would make sense to get it reviewed there first before submission here? If not for this time, then next time.

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md:1]: Good point, simplified it by `Bundle, test and deploy Gatekeeper policies as OCI image`

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md]: Done

Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md]: Done

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md]: Done.

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md]: Done.

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md]: Done.

**Comment by mathieu-benoit** [comment]
Thanks @vsoch for carefully reviewing this blog post, much appreciated! And glad you liked it!

All your comments and suggestions are now taken into account, let me know if there is anything else. Thanks!

As for your question/statement:

> I know Google has documentation standards and review - perhaps it would make sense to get it reviewed there first before submission here?

All good in that aspect, 2 of my peers already gave a quick review in the context of this blog post _from the community for the community_. If you have any specific concern or feedback, I am more than happy to take them into account.

**Comment by mathieu-benoit** [comment]
Out of curiosity @vsoch, anything I missed from your review to get your approval on this PR? Thanks!

**Review by vsoch — APPROVED** [review]
Approved! I mostly just needed a ping - sorry for the delay!

**Comment by mathieu-benoit** [comment]
No worries at all, thanks @vsoch!

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (SteveLasker) [inline docs/blog/gatekeeper-policies-as-oci-image.md:143]: pushing without an artifactType or --config mediaType means the gatekeeper policy has no type. Or, the default of unknown

Suggest providing some type by adding:
`--config /dev/null:application/vnd.cncf.gatekeeper.v1`

**Review by SteveLasker — COMMENTED** [review]
Great article @mathieu-benoit 
I had one comment about the "type" as it's known within the registry. 
Not passing a type means it's "unknown" by the registry, and anything trying to scan contents in the registry.

File: docs/blog/gatekeeper-policies-as-oci-image.md
Comment (SteveLasker) [reply docs/blog/gatekeeper-policies-as-oci-image.md:143]: cc: @garethr for the actual mediaType as I know he'd done some work earlier.
Another suggestion would be to include the oras-go library into `gator` so you could do a `gator push|pull` directly. In this case, you could embed the appropriate `artifactType` directly into the gather cli.

Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md:143]: Thanks for your review @SteveLasker!
Yes indeed the `config.mediaType` is ``:
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.manifest.v1+json",
   "config": {
      "mediaType": "application/vnd.unknown.config.v1+json",
      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "size": 2
   },
   "layers": [
      {
         "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
         "digest": "sha256:325baa2355262c1cb5d8fb274759405f5c4b26e2110dfc22bbf9565740757938",
         "size": 1027,
         "annotations": {
            "io.deis.oras.content.digest": "sha256:6d2f0cfca84142481087d9566dd9e3a77a69d3e1a3754595c1615d09348487d8",
            "io.deis.oras.content.unpack": "true",
            "org.opencontainers.image.title": "policies"
         }
      }
   ],
   "annotations": {
      "org.opencontainers.image.created": "2023-02-08T17:11:52Z"
   }
}
```

Where this `mediaType` `application/vnd.cncf.gatekeeper.v1` is coming from and defined? Is it like a feature request you are suggesting for the Gatekeeper/`gator` project? Not seeing anything in that regard in there: https://open-policy-agent.github.io/gatekeeper/website/docs/gator/#bundling-policy-into-oci-artifacts.

Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md:143]: Regarding your other point about `oras-go` in `gator`, that's already the case: https://github.com/open-policy-agent/gatekeeper/pull/2398. That's what `gator test|expand --image` is leveraging, FYI.

Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md:143]: Also, @SteveLasker, is it blocking this PR to get your approval to be shipped in `main`? Maybe this should be a separate discussion in Slack or also, I'm more than happy to file an issue in Gatekeeper/`gator` repo to have this discussion there (depending on the answers of my questions above). Thanks!

Comment (SteveLasker) [reply docs/blog/gatekeeper-policies-as-oci-image.md:143]: Thanks, @mathieu-benoit 
The mediaType usage is outlined here: [OCI Artifacts: Defining a Unique Artifact Type](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#defining-a-unique-artifact-type)

As for which `config.mediaType` to use, that should be defined by the OPA project, and that's why I was pinging @garethr, as I remember him doing so in a prior experiment. 

I wouldn't say it's blocking, rather the article does a great job and users tend to copy/paste examples, so I was hoping to add this clarity so everything in the registry didn't look "unknown". 
Here's another example [Visualizing Artifacts](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#visualizing-artifacts), which is also supported by Docker Hub. See [Announcing Docker Hub OCI Artifacts Support](https://www.docker.com/blog/announcing-docker-hub-oci-artifacts-support/), with the visualization section title: **Viewing your Helm chart and using filters**

I see the additional gator support, which is awesome. It just looks like this `config.mediaType` was copied over as unknown, rather than identifying a well-known type.

Comment (mathieu-benoit) [reply docs/blog/gatekeeper-policies-as-oci-image.md:143]: Gotcha, thanks for clarifying. 

I'll go ahead and resolve this discussion to unblock the review/shipment of this blog post. Hope that's ok?

As soon as the `config.mediaType` value to use is defined by the OPA/Gatekeeper project, it will be my pleasure to update that blog post accordingly.

Thanks!

**Comment by FeynmanZhou** [comment]
Now all comments are resolved by @mathieu-benoit . We are ready to merge it into main.

**Comment by mathieu-benoit** [comment]
Woot! Woot! Thanks everyone for your review!

**Pull request merged**
